syntax = "proto3";

package Serialize;

message StopNumber {
    int32 stop_number = 1;
    string stop_name = 2;
}

message BusResponse {
    string bus_name = 1;
    int32 route_length = 2;
    double curvature = 3;
    uint64 stop_count = 4;
    uint64 unique_stop_count = 5;

    // extra
    repeated string stops = 6;
    bool is_roundtrip = 7;
}

message GeoCoords {
    double latitude = 1;
    double longitude = 2;
}

message StopResponse {
    string stop_name = 1;
    repeated string buses = 2;

    // extra
    GeoCoords geo_coords = 3;
}

// 2 part
message RouteElement {
    enum RouteElementType {
        WAIT_ELEMENT = 0;
        BUS_ELEMENT = 1;
    }
    RouteElementType type = 1;

    // WAIT_ELEMENT
    int32 stop_number = 2;

    // BUS_ELEMENT
    double time = 3;
    string bus = 4;
    int32 span_count = 5;
    int32 from_idx = 6;
    int32 to_idx = 7;
}

message RouteResponse {
    int32 from_stop_number = 1;
    int32 to_stop_number = 2;
    repeated RouteElement route_elements = 3;
    double total_time = 4;
}

// 3 part
message Point {
    double x = 1;
    double y = 2;
}

message Color {
    enum ColorType {
        STR = 0;
        RGB = 1;
        RGBA = 2;
    }

    ColorType type = 1;
    string str = 2;
    int32 red = 3;
    int32 green = 4;
    int32 blue = 5;
    double alpha = 6;
}

message RenderSettings {
    double width = 1;
    double height = 2;
    double padding = 3;
    double stop_radius = 4;
    double line_width = 5;
    int32 stop_label_font_size = 6;
    Point stop_label_offset = 7;
    Color underlayer_color = 8;
    double underlayer_width = 9;
    repeated Color color_palette = 10;
    int32 bus_label_font_size = 11;
    Point bus_label_offset = 12;
    repeated string layers = 13;
    double outer_margin = 14;
}

message TransportCatalog  {
    // 1 part
    repeated BusResponse bus_responses = 1;
    repeated StopResponse stop_responses = 2;

    // 2 part
    int32 wait_time = 3;
    repeated StopNumber stop_numbers = 4;
    repeated RouteResponse route_responses = 5;

    // 3 part
    RenderSettings render_settings = 6;
}
